src/distfile.py

```diff
@@ -101,50 +110,53 @@ def _fetch(self):
             except FileNotFoundError as e:
                 log.warning(e)
 
-            url = random.sample(self._urls, 1)[0]
+            url = io.random.sample(self._urls, 1)[0]
             log.info('FETCH %s', url)
             try:
                 util.make_parent_dir(self._pathname)
-                with util.unsafe_fetch(url) as fin, open(self._pathname, 'wb') as fout:
-                    shutil.copyfileobj(fin, fout)
+                with util.unsafe_fetch(url, io.urlopen) as fin, self._pathname.open('wb') as fout:
+                    copyfileobj(fin, fout)
             except ConnectionResetError as e:
                 log.warning(e)
-            except urllib.error.URLError as e:
+            except URLError as e:
                 log.warning(e)
         raise Exception('Failed to fetch %s' % self._name)
 
     def extract(self, target):
+        io = self._io
         target = self._extract_unpatched(target)
         # Apply patches.
         for patch in self._patches:
-            self._apply_patch(patch, target)
+            self._apply_patch(patch, target, io.subprocess)
         # Add markers to sources that depend on unsafe string sources.
         for filename in self._unsafe_string_sources:
-          path = os.path.join(target, filename)
-          with open(path, 'rb') as fin, open(path + '.new', 'wb') as fout:
-            fout.write(bytes('#define _CLOUDLIBC_UNSAFE_STRING_FUNCTIONS\n',
-                             encoding='ASCII'))
-            fout.write(fin.read())
-            os.rename(path + '.new', path)
+            path = target / filename
+            with path.open('rb') as fin, (path + '.new').open('wb') as fout:
+                fout.write(
+                    bytes('#define _CLOUDLIBC_UNSAFE_STRING_FUNCTIONS\n',
+                          encoding='ASCII'))
+                fout.write(fin.read())
```

Comment:
Now that I'm reading my own code: should we replace this with `shutil.copyfileobj(fin, fout)`?